File: applications/rtkadmmwavelets/rtkadmmwavelets.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkadmmwavelets_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

#include "rtkProjectionsReader.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkConstantImageSource.h"
#include "rtkDisplacedDetectorImageFilter.h"
#include "rtkADMMWaveletsConeBeamReconstructionFilter.h"

int
main(int argc, char * argv[])
{
  GGO(rtkadmmwavelets, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
#endif

  //////////////////////////////////////////////////////////////////////////////////////////
  // Read all the inputs
  //////////////////////////////////////////////////////////////////////////////////////////

  // Projections reader
  using projectionsReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto projectionsReader = projectionsReaderType::New();
  rtk::SetProjectionsReaderFromGgo<projectionsReaderType, args_info_rtkadmmwavelets>(projectionsReader, args_info);

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Displaced detector weighting
  using DDFType = rtk::DisplacedDetectorImageFilter<OutputImageType>;
  auto ddf = DDFType::New();
  ddf->SetInput(projectionsReader->GetOutput());
  ddf->SetGeometry(geometry);

  // Create input: either an existing volume read from a file or a blank image
  itk::ImageSource<OutputImageType>::Pointer inputFilter;
  if (args_info.input_given)
  {
    // Read an existing image to initialize the volume
    using InputReaderType = itk::ImageFileReader<OutputImageType>;
    auto inputReader = InputReaderType::New();
    inputReader->SetFileName(args_info.input_arg);
    inputFilter = inputReader;
  }
  else
  {
    // Create new empty volume
    using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
    auto constantImageSource = ConstantImageSourceType::New();
    rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkadmmwavelets>(constantImageSource,
                                                                                           args_info);
    inputFilter = constantImageSource;
  }

  //////////////////////////////////////////////////////////////////////////////////////////
  // Setup the ADMM filter and run it
  //////////////////////////////////////////////////////////////////////////////////////////

  // Set the reconstruction filter
  using ADMM_Wavelets_FilterType = rtk::ADMMWaveletsConeBeamReconstructionFilter<OutputImageType>;
  auto admmFilter = ADMM_Wavelets_FilterType::New();

  // Set the forward and back projection filters to be used inside admmFilter
  SetForwardProjectionFromGgo(args_info, admmFilter.GetPointer());
  SetBackProjectionFromGgo(args_info, admmFilter.GetPointer());

  // Set the geometry and interpolation weights
  admmFilter->SetGeometry(geometry);

  // Set all numerical parameters
  admmFilter->SetCG_iterations(args_info.CGiter_arg);
  admmFilter->SetAL_iterations(args_info.niterations_arg);
  admmFilter->SetAlpha(args_info.alpha_arg);
  admmFilter->SetBeta(args_info.beta_arg);
  admmFilter->SetNumberOfLevels(args_info.levels_arg);
  admmFilter->SetOrder(args_info.order_arg);

  // Set the inputs of the ADMM filter
  admmFilter->SetInput(0, inputFilter->GetOutput());
  admmFilter->SetInput(1, projectionsReader->GetOutput());

  admmFilter->SetDisableDisplacedDetectorFilter(args_info.nodisplaced_flag);

  REPORT_ITERATIONS(admmFilter, ADMM_Wavelets_FilterType, OutputImageType);

  TRY_AND_EXIT_ON_ITK_EXCEPTION(admmFilter->Update())

  // Write the output
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(admmFilter->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkadmmwavelets/rtkadmmwavelets.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Performs an iterative 3D reconstruction with Daubechies wavelets regularization.",
    )

    # GGO options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--niterations", "-n", help="Number of iterations", type=int, default=1
    )
    parser.add_argument(
        "--alpha", help="Regularization parameter", type=float, default=0.1
    )
    parser.add_argument(
        "--beta",
        help="Augmented Lagrangian constraint multiplier",
        type=float,
        default=1,
    )
    parser.add_argument(
        "--CGiter",
        help="Number of nested iterations of conjugate gradient",
        type=int,
        default=5,
    )
    parser.add_argument(
        "--order", help="The order of the Daubechies wavelets", type=int, default=3
    )
    parser.add_argument(
        "--levels",
        help="The number of decomposition levels in the wavelets transform",
        type=int,
        default=5,
    )
    parser.add_argument("--input", "-i", help="Input volume", type=str)
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )

    # Add additional RTK output image options if needed
    rtk.add_rtk3Doutputimage_group(parser)
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtkiterations_group(parser)
    rtk.add_rtkprojectors_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):

    # Define output pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    projectionsReader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(projectionsReader, args_info)

    if args_info.verbose:
        print(f"Reading projections...")
    projectionsReader.Update()

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")

    geometry = rtk.read_geometry(args_info.geometry)

    # Create input volume: either an existing volume read from a file or a blank image
    if args_info.input:
        inputFilter = itk.imread(args_info.input)
    else:
        constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
        rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)
        inputFilter = constantImageSource

    # Setup the ADMM wavelets reconstruction filter
    if hasattr(itk, "CudaImage"):
        CudaOutputImageType = itk.CudaImage[OutputPixelType, Dimension]
        admmFilter = rtk.ADMMWaveletsConeBeamReconstructionFilter[
            CudaOutputImageType
        ].New()
        admmFilter.SetInput(0, itk.cuda_image_from_image(inputFilter.GetOutput()))
        admmFilter.SetInput(1, itk.cuda_image_from_image(projectionsReader.GetOutput()))
    else:
        admmFilter = rtk.ADMMWaveletsConeBeamReconstructionFilter[OutputImageType].New()
        admmFilter.SetInput(0, inputFilter.GetOutput())
        admmFilter.SetInput(1, projectionsReader.GetOutput())

    # Set the forward and back projection filters to be used inside admmFilte
    rtk.SetForwardProjectionFromArgParse(args_info, admmFilter)
    rtk.SetBackProjectionFromArgParse(args_info, admmFilter)

    # Set the geometry and interpolation weights
    admmFilter.SetGeometry(geometry)

    # Set numerical parameters
    admmFilter.SetCG_iterations(args_info.CGiter)
    admmFilter.SetAL_iterations(args_info.niterations)
    admmFilter.SetAlpha(args_info.alpha)
    admmFilter.SetBeta(args_info.beta)
    admmFilter.SetNumberOfLevels(args_info.levels)
    admmFilter.SetOrder(args_info.order)

    admmFilter.SetDisableDisplacedDetectorFilter(args_info.nodisplaced)

    rtk.SetIterationsReportFromArgParse(args_info, admmFilter)

    admmFilter.Update()

    writer = itk.ImageFileWriter[OutputImageType].New()
    writer.SetFileName(args_info.output)
    writer.SetInput(admmFilter.GetOutput())
    writer.Update()

    if args_info.verbose:
        print("Processing completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
